In [1]:
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.svm import SVC

X_train = np.load('../../data/X_train.npy')
X_test = np.load('../../data/X_test.npy')
y_train = np.load('../../data/Y_train.npy')
y_test = np.load('../../data/Y_test.npy')

print(X_train.shape)

(292, 13)


# **Classificação**

Nosso objetivo aqui é montar um modelo de classificação, o qual classificará os slientes em confiáveis e não confiáveis.

A variável **default** é utilizada como proxy para o risco de inadimplência, permitindo classificar clientes em confiáveis e não confiáveis com base em padrões históricos de comportamento financeiro.

Testaremos dois modelos diferentes, sendo eles:
* **Randon Forest:** combina múltiplas árvores de decisão independentes, utilizando um sistema de votação para reduzir erros individuais e aumentar a robustez do modelo.

* **Gradient Boosting:** constrói árvores de forma sequencial, onde cada novo modelo busca corrigir os erros dos anteriores, resultando em maior capacidade preditiva.

* **Support Vector Machine:** O Support Vector Machine busca encontrar a melhor fronteira de separação entre classes, maximizando a margem entre clientes confiáveis e não confiáveis.

## **Random Forest**

In [2]:
#modelo
model_rf = RandomForestClassifier()
model_rf.fit(X_train, y_train)

#previsao
y_pred_rf = model_rf.predict(X_test)

#avaliação
print("Resultados modelo Random Forest")
report = classification_report(y_test, y_pred_rf, output_dict=True)

df_report = pd.DataFrame(report).transpose()
df_report

Resultados modelo Random Forest


,precision,recall,f1-score,support
0.0,0.914286,0.621359,0.739884,103.000000
1.0,0.442857,0.837838,0.579439,37.000000
accuracy,0.678571,0.678571,0.678571,0.678571
macro avg,0.678571,0.729599,0.659662,140.000000
weighted avg,0.789694,0.678571,0.697481,140.000000


## **Gradient Boosting**

In [3]:
# modelo
model_gb = GradientBoostingClassifier()
model_gb.fit(X_train, y_train)

# previsão
y_pred_gb = model_gb.predict(X_test)

# avaliação
print("Resultados modelo Gradient Boosting")

report_gb = classification_report(y_test, y_pred_gb, output_dict=True)
df_report_gb = pd.DataFrame(report_gb).transpose()

df_report_gb

Resultados modelo Gradient Boosting


,precision,recall,f1-score,support
0.0,0.906667,0.660194,0.764045,103.0
1.0,0.461538,0.810811,0.588235,37.0
accuracy,0.700000,0.700000,0.700000,0.7
macro avg,0.684103,0.735502,0.676140,140.0
weighted avg,0.789026,0.700000,0.717581,140.0


# **Support Vector Machine**

In [4]:
# modelo
model_svm = SVC(probability=True) # probability=True permite extrair probabilidades depois se precisar
model_svm.fit(X_train, y_train)

# previsão
y_pred_svm = model_svm.predict(X_test)

# avaliação
print("Resultados modelo SVM")

report_svm = classification_report(y_test, y_pred_svm, output_dict=True)
df_report_svm = pd.DataFrame(report_svm).transpose()

df_report_svm

Resultados modelo SVM


,precision,recall,f1-score,support
0.0,0.943662,0.650485,0.770115,103.000000
1.0,0.478261,0.891892,0.622642,37.000000
accuracy,0.714286,0.714286,0.714286,0.714286
macro avg,0.710961,0.771189,0.696378,140.000000
weighted avg,0.820663,0.714286,0.731140,140.000000


Agora que temos nosso modelos e suas respectivas métricas de avaliação, precisamos entender cada uma e, assim, decidir qual modelo usar. 
* De modo geral, a **precisão** mede a qualidade da nossa decisão (não queremos ser injustos com bons clientes), enquanto o **recall** mede o nosso poder de detecção (não queremos deixar caloteiros passarem despercebidos).


## **Escolha**

Considerando os resultados obtidos, observamos que o **Support Vector Machine (SVM)** apresentou o desempenho inferior na classificação da **Categoria 1** (clientes não confiáveis/inadimplentes). Devido a essa baixa performance em identificar o grupo de risco, o modelo foi descartado da seleção final.

A disputa concentrou-se entre o **Random Forest** e o **Gradient Boosting**. Ambos apresentaram níveis satisfatórios e muito similares de acurácia e precisão. No entanto, o Gradient Boosting destacou-se por obter um **Recall** superior para a categoria de risco (Categoria 1).

Enquanto o Random Forest identificou apenas **59%** dos casos reais de inadimplência, o Gradient Boosting alcançou **67%**. Em um cenário de análise de dados bancários, o objetivo primordial é mitigar o risco de crédito, evitando a concessão de empréstimos a perfis não confiáveis. Portanto, priorizamos o modelo com maior capacidade de detecção (maior Recall).

Com base nesses critérios, o **Gradient Boosting** foi o modelo escolhido para a implementação.